### PR TITLE
foreman as inv plugin

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

needs testing

##### ISSUE TYPE

 - Feature Pull Request
##### COMPONENT NAME

foreman inventory plugin
##### ANSIBLE VERSION

```
2.4
```

### CODE REVIEW

**Comment by bcoca** [comment]
cc @dLobatog, @Fobhep, @ehelms, @clasohm, @ericzolf

File: lib/ansible/plugins/inventory/foreman.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+
+class InventoryModule(BaseInventoryPlugin):
+    ''' Host inventory parser for ansible using foreman as source. '''
+
+    NAME = 'foreman'
+
+    def __init__(self):
+
+        super(InventoryModule, self).__ini__()
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/foreman.py]: `__init__()`

File: lib/ansible/plugins/inventory/foreman.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
+# This file is part of Ansible,
+# Copyright (C) 2016 Guido Günther <agx@sigxcpu.org>,
+#                    Daniel Lobato Garcia <dlobatog@redhat.com>
+#                2017 Ansible by Red Hat, inc
+#
+# This script is free software: you can redistribute it and/or modify
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/foreman.py]: Do we use the new copyright for plugins?

File: lib/ansible/plugins/inventory/foreman.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/foreman.py]: good catch

File: lib/ansible/plugins/inventory/foreman.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
+# Copyright (C) 2016 Guido Günther <agx@sigxcpu.org>, Daniel Lobato Garcia <dlobatog@redhat.com>
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: forman
```
Comment (dLobatog) [inline lib/ansible/plugins/inventory/foreman.py]: typo, foreman

File: lib/ansible/plugins/inventory/foreman.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    plugin_type: inventory
+    short_description: foreman inventory source
+    description:
+        - Get inventory hosts from the foreman service.
+    options:
+      url:
+        description: TODO
```
Comment (dLobatog) [inline lib/ansible/plugins/inventory/foreman.py]: You can grab some descriptions for these options from https://github.com/ansible/ansible/blob/devel/contrib/inventory/foreman.ini

File: lib/ansible/plugins/inventory/foreman.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      user:
+        description: TODO
+      password:
+        description: TODO
+      ssl_verify:
+        description: TODO
+      group_prefix:
+        description: TODO
```
Comment (dLobatog) [inline lib/ansible/plugins/inventory/foreman.py]: `want_facts`, `want_hostcollections`, `rich_params` are missing. Also there is no option to change the cache max age?

File: lib/ansible/plugins/inventory/foreman.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    ''' Host inventory parser for ansible using foreman as source. '''
+
+    NAME = 'foreman'
+
+    def __init__(self):
+
+        super(InventoryModule, self).__init__()
+
+        # from config
+        self.foreman_url = None
```
Comment (dLobatog) [inline lib/ansible/plugins/inventory/foreman.py:75]: Is there a way to raise errors in Ansible when this attribute is None?

**Review by dLobatog — CHANGES_REQUESTED** [review]
Thanks @bcoca , it looks good to me in general, can you add the Host collections part? Seems like the inventory was updated with a couple of commits after this was submitted.

File: lib/ansible/plugins/inventory/foreman.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/foreman.py]: will add those, cache is something I want to implement at a 'global' level for all inventory plugins, i'll add cache control options once that is done

File: lib/ansible/plugins/inventory/foreman.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/foreman.py:75]: I can `raise AnsibleOptionsError`

File: lib/ansible/plugins/inventory/foreman.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+    def verify_file(self, path):
+
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith('.foreman.yaml') or path.endswith('.foreman.yml'):
+                valid = True
+        return valid
+
+    def _get_session(self):
```
Comment (webknjaz) [inline lib/ansible/plugins/inventory/foreman.py:89]: Why not turn this into `@property` which it effectively is?

File: lib/ansible/plugins/inventory/foreman.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+                        break
+
+                    # get next page
+                    params['page'] += 1
+
+            self.cache[url] = results
+
+        return self.cache[url]
+
+    def _get_hosts(self):
```
Comment (webknjaz) [inline lib/ansible/plugins/inventory/foreman.py]: this is a property as well

File: lib/ansible/plugins/inventory/foreman.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        ret = self._get_facts_by_id(host['id'])
+        if len(ret.values()) == 0:
+            facts = {}
+        elif len(ret.values()) == 1:
+            facts = list(ret.values())[0]
+        else:
+            raise ValueError("More than one set of facts returned for '%s'" % host)
+        return facts
+
+    def to_safe(self, word):
```
Comment (webknjaz) [inline lib/ansible/plugins/inventory/foreman.py:168]: It's a `@staticmethod`/pure function

File: lib/ansible/plugins/inventory/foreman.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        '''Converts 'bad' characters in a string to underscores so they can be used as Ansible groups
```
Comment (webknjaz) [inline lib/ansible/plugins/inventory/foreman.py:169]: Do we have PEP257 linter in CI? It should've failed here.

File: lib/ansible/plugins/inventory/foreman.py (hunk 7 above)
Comment (bcoca) [reply lib/ansible/plugins/inventory/foreman.py:89]: mostly this is a copy of the original, script, i tried to modify as little as possible to adapt it into a plugin

**Comment by mattclay** [comment]
Sanity failure is valid. The integration test failure is unrelated.

**Comment by bcoca** [comment]
bot_status

**Pull request merged**